File: backend/src/utils/ludo_engine.py
import random

COLORS = ("red", "green", "yellow", "blue")
START_OFFSETS = {"red": 0, "green": 13, "yellow": 26, "blue": 39}
SAFE_SQUARES = {0, 8, 13, 21, 26, 34, 39, 47}

BASE_POS = -1
TRACK_END = 51
HOME_DONE = 58
TOKENS_PER_PLAYER = 4


def roll_dice() -> int:
    return random.randint(1, 6)


def initial_state(participants: list[dict]) -> dict:
    return {
        "tokens": {p["color"]: [BASE_POS] * TOKENS_PER_PLAYER for p in participants},
        "participants": participants,
    }


def legal_moves(tokens: list[int], dice: int) -> list[dict]:
    moves = []
    for index, pos in enumerate(tokens):
        if pos == HOME_DONE:
            continue
        if pos == BASE_POS:
            if dice == 6:
                moves.append({"token_index": index, "to_pos": 0})
            continue
        target = pos + dice
        if target <= HOME_DONE:
            moves.append({"token_index": index, "to_pos": target})
    return moves


def abs_square(color: str, pos: int) -> int | None:
    if pos < 0 or pos > TRACK_END:
        return None
    return (START_OFFSETS[color] + pos) % 52


def apply_move(state: dict, color: str, token_index: int, dice: int) -> dict:
    tokens = state["tokens"][color]
    current = tokens[token_index]

    if current == BASE_POS:
        target = 0
    else:
        target = current + dice
        if target > HOME_DONE:
            raise ValueError("Move overshoots home")

    tokens[token_index] = target

    captured = False
    absolute = abs_square(color, target)
    if absolute is not None and absolute not in SAFE_SQUARES:
        for other in state["tokens"]:
            if other == color:
                continue
            offset = START_OFFSETS[other]
            for i, opp_pos in enumerate(state["tokens"][other]):
                if opp_pos < 0 or opp_pos > TRACK_END:
                    continue
                if (offset + opp_pos) % 52 == absolute:
                    state["tokens"][other][i] = BASE_POS
                    captured = True

    return {
        "captured": captured,
        "reached_home": target == HOME_DONE,
        "finished": all(p == HOME_DONE for p in tokens),
    }


def progress(color_tokens: list[int]) -> int:
    return sum(1 for p in color_tokens if p != BASE_POS)


def is_color_finished(state: dict, color: str) -> bool:
    return all(p == HOME_DONE for p in state["tokens"][color])


def next_turn(participants: list[dict], current_color: str, state: dict) -> str:
    colors = [p["color"] for p in participants]
    index = colors.index(current_color)
    for step in range(1, len(colors) + 1):
        candidate = colors[(index + step) % len(colors)]
        if not is_color_finished(state, candidate):
            return candidate
    return current_color

File: flutter_app/lib/game/ludo_engine.dart
import 'board_geometry.dart';

/// Pure Ludo rules engine shared by local play and online state rendering.
class LudoEngine {
  LudoEngine._();

  static const int basePos = -1;
  static const int trackEnd = 51;
  static const int laneStart = 52;
  static const int homeDone = 57;

  static List<int> initialTokens() => [basePos, basePos, basePos, basePos];

  /// Target position per token index; null where the token cannot move.
  static List<int?> legalTargets(List<int> tokens, int dice) {
    final targets = List<int?>.filled(tokens.length, null);
    for (var i = 0; i < tokens.length; i++) {
      final pos = tokens[i];
      if (pos == homeDone) continue;
      if (pos == basePos) {
        if (dice == 6) targets[i] = 0;
        continue;
      }
      final target = pos + dice;
      if (target <= homeDone) targets[i] = target;
    }
    return targets;
  }

  static bool hasAnyMove(List<int> tokens, int dice) =>
      legalTargets(tokens, dice).any((t) => t != null);

  static MoveResult applyMove(
    Map<String, List<int>> allTokens,
    String color,
    int tokenIndex,
    int dice,
  ) {
    final tokens = allTokens[color]!;
    final current = tokens[tokenIndex];

    final target = current == basePos ? 0 : current + dice;
    if (target > homeDone) throw StateError('Move overshoots home');
    tokens[tokenIndex] = target;

    final capturedColors = <String>[];
    final victims = <CapturedToken>[];
    var captured = false;

    final abs = target <= trackEnd ? BoardGeometry.absoluteSquare(color, target) : -1;
    if (abs >= 0 && !BoardGeometry.safeSquares.contains(abs)) {
      for (final other in BoardGeometry.colors) {
        if (other == color || !allTokens.containsKey(other)) continue;
        final offset = BoardGeometry.startOffsets[other]!;
        final oppTokens = allTokens[other]!;
        for (var i = 0; i < oppTokens.length; i++) {
          final p = oppTokens[i];
          if (p < 0 || p > trackEnd) continue;
          if ((offset + p) % 52 == abs) {
            oppTokens[i] = basePos;
            captured = true;
            victims.add(CapturedToken(color: other, tokenIndex: i, fromPos: p));
            if (!capturedColors.contains(other)) capturedColors.add(other);
          }
        }
      }
    }

    return MoveResult(
      from: current,
      to: target,
      captured: captured,
      capturedColors: capturedColors,
      victims: victims,
      reachedHome: target == homeDone,
      finished: tokens.every((p) => p == homeDone),
    );
  }

  static bool isFinished(List<int> tokens) => tokens.every((p) => p == homeDone);

  /// Capture/finish resolution for a token that has ALREADY been moved
  /// (animated step by step) to [finalPos].
  static MoveResult resolveAfterMove(
    Map<String, List<int>> allTokens,
    String color,
    int finalPos,
  ) {
    final capturedColors = <String>[];
    final victims = <CapturedToken>[];
    var captured = false;

    final abs =
        finalPos <= trackEnd ? BoardGeometry.absoluteSquare(color, finalPos) : -1;
    if (abs >= 0 && !BoardGeometry.safeSquares.contains(abs)) {
      for (final other in BoardGeometry.colors) {
        if (other == color || !allTokens.containsKey(other)) continue;
        final offset = BoardGeometry.startOffsets[other]!;
        final oppTokens = allTokens[other]!;
        for (var i = 0; i < oppTokens.length; i++) {
          final p = oppTokens[i];
          if (p < 0 || p > trackEnd) continue;
          if ((offset + p) % 52 == abs) {
            oppTokens[i] = basePos;
            captured = true;
            victims.add(CapturedToken(color: other, tokenIndex: i, fromPos: p));
            if (!capturedColors.contains(other)) capturedColors.add(other);
          }
        }
      }
    }

    return MoveResult(
      from: -2,
      to: finalPos,
      captured: captured,
      capturedColors: capturedColors,
      victims: victims,
      reachedHome: finalPos == homeDone,
      finished: allTokens[color]!.every((p) => p == homeDone),
    );
  }

  static int tokensHome(List<int> tokens) =>
      tokens.where((p) => p == homeDone).length;

  static int progressOf(List<int> tokens) => tokens
      .where((p) => p != basePos)
      .fold(0, (sum, p) => sum + (p == homeDone ? trackEnd + 6 : p));

  static String nextColor(
    List<String> order,
    String current,
    Map<String, List<int>> tokens,
  ) {
    final index = order.indexOf(current);
    for (var step = 1; step <= order.length; step++) {
      final candidate = order[(index + step) % order.length];
      if (!isFinished(tokens[candidate]!)) return candidate;
    }
    return current;
  }
}

class CapturedToken {
  const CapturedToken({
    required this.color,
    required this.tokenIndex,
    required this.fromPos,
  });

  final String color;
  final int tokenIndex;

  /// Ring position the victim occupied when it was cut.
  final int fromPos;
}

/// Visual effect kinds for the presentation layer.
enum FxKind { flame, sparkle }

/// One board square where an effect should play.
class FxSpot {
  const FxSpot({
    required this.color,
    required this.tokenIndex,
    required this.pos,
  });

  final String color;
  final int tokenIndex;
  final int pos;
}

/// A numbered visual event (kill flames / home sparkles) for the UI.
class BoardFx {
  const BoardFx({required this.id, required this.kind, required this.spots});

  final int id;
  final FxKind kind;
  final List<FxSpot> spots;
}

class MoveResult {
  MoveResult({
    required this.from,
    required this.to,
    required this.captured,
    required this.capturedColors,
    this.victims = const [],
    required this.reachedHome,
    required this.finished,
  });

  final int from;
  final int to;
  final bool captured;
  final List<String> capturedColors;
  final List<CapturedToken> victims;
  final bool reachedHome;
  final bool finished;
}
